core/src/snapshot.rs

```diff
@@ -182,6 +182,20 @@ pub fn try_read_snapshot(
             });
         }
     }
+    // If last block hash is different it might mean that snapshot was crated for soft-fork block so just drop changes made by this block
+    if snapshot_height > 0 {
+        let kura_block_hash = NonZeroUsize::new(snapshot_height)
+            .and_then(|height| kura.get_block_hash(height))
+            .expect("Kura has height at least as large as state height");
+        let snapshot_block_hash = state_view.block_hashes[snapshot_height - 1];
+        if kura_block_hash != snapshot_block_hash {
+            iroha_logger::warn!(
+                "Snapshot has incorrect latest block hash, discarding changes made by this block"
+            );
+            state.block_and_revert().commit();
+        }
+    }
```

Comment:
This could be combined with the loop just behind, reducing extra duplication (and confusion):

```rust
for height in 1..=snapshot_height {
    let kura_block_hash = NonZeroUsize::new(height)
        .and_then(|height| kura.get_block_hash(height))
        .expect("Kura has height at least as large as state height");
    let snapshot_block_hash = state_view.block_hashes[height - 1];
    if kura_block_hash != snapshot_block_hash {
        if height == snapshot_height {
            // If last block hash is different it might mean that snapshot was crated for soft-fork block so just drop changes made by this block
            iroha_logger::warn!(
                "Snapshot has incorrect latest block hash, discarding changes made by this block"
            );
            state.block_and_revert().commit();
        } else {
            return Err(TryReadError::MismatchedHash {
                height,
                snapshot_block_hash,
                kura_block_hash,
            });
        }
    }
}
```

Reply:
My reasoning was to exclude unnecessary if check out of loop, but since it's one off function it doesn't matter that much.